cpp/include/cuspatial/experimental/detail/is_point_in_polygon_kernel.cuh

```diff
@@ -0,0 +1,102 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <cuspatial/traits.hpp>
+
+namespace cuspatial {
+namespace detail {
+
+/**
+ * @brief Kernel to test if a point is inside a polygon.
+ *
+ * Implemented based on Eric Haines's crossings-multiply algorithm:
+ * See "Crossings test" section of http://erich.realtimerendering.com/ptinpoly/
+ * The improvement in addenda is also addopted to remove divisions in this kernel.
+ *
+ * TODO: the ultimate goal of refactoring this as independent function is to remove
+ * src/utility/point_in_polygon.cuh and its usage in quadtree_point_in_polygon.cu. It isn't
+ * possible today without further work to refactor quadtree_point_in_polygon into header only
+ * API.
+ */
+template <class Cart2d,
+          class OffsetType,
+          class OffsetIterator,
+          class Cart2dIt,
+          class OffsetItDiffType = typename std::iterator_traits<OffsetIterator>::difference_type,
+          class Cart2dItDiffType = typename std::iterator_traits<Cart2dIt>::difference_type>
+__device__ inline bool is_point_in_polygon(Cart2d const& test_point,
+                                           OffsetType poly_begin,
+                                           OffsetType poly_end,
+                                           OffsetIterator ring_offsets_first,
+                                           OffsetItDiffType const& num_rings,
+                                           Cart2dIt poly_points_first,
+                                           Cart2dItDiffType const& num_poly_points)
+{
+  using T = iterator_vec_base_type<Cart2dIt>;
+
+  bool point_is_within = false;
+  bool is_colinear     = false;
+  // for each ring
+  for (auto ring_idx = poly_begin; ring_idx < poly_end; ring_idx++) {
+    int32_t ring_idx_next = ring_idx + 1;
+    int32_t ring_begin    = ring_offsets_first[ring_idx];
+    int32_t ring_end =
+      (ring_idx_next < num_rings) ? ring_offsets_first[ring_idx_next] : num_poly_points;
+
+    Cart2d b     = poly_points_first[ring_end - 1];
+    bool y0_flag = b.y > test_point.y;
+    bool y1_flag;
+    // for each line segment, including the segment between the last and first vertex
+    for (auto point_idx = ring_begin; point_idx < ring_end; point_idx++) {
+      Cart2d const a = poly_points_first[point_idx];
+      T run          = b.x - a.x;
+      T rise         = b.y - a.y;
+
+      // Points on the line segment are the same, so intersection is impossible.
+      // This is possible because we allow closed or unclosed polygons.
+      if (run == 0.0 && rise == 0.0) continue;
+
+      T rise_to_point = test_point.y - a.y;
+
+      // colinearity test
+      T run_to_point = test_point.x - a.x;
+      is_colinear    = (run * rise_to_point - run_to_point * rise) == 0;
```

Comment:
Thanks for those corner cases. I'll be sure to check them against GeoPandas and add an epsilon if they handle those cases better than us.

Reply:
GeoPandas has inconsistent results along the boundary as well. I'm not sure how much time we should spend chasing floating point error, what do you recommend?

I wrote the following test for GeoPandas and cuSpatial:

```
@pytest.mark.parametrize(
    "point, polygon, expects",
    [
        # unique failure cases identified by @mharris
        [Point([0.6, 0.06]), Polygon([[0, 0], [10, 1], [1, 1], [0, 0]]), False],
        [Point([3.3, 1.1]), Polygon([[6, 2], [3, 1], [3, 4], [6, 2]]), True],
    ],
)
def test_float_precision_limits(point, polygon, expects):
    gpdpoint = gpd.GeoSeries(point)
    gpdpolygon = gpd.GeoSeries(polygon)
    point = cuspatial.from_geopandas(gpdpoint)
    polygon = cuspatial.from_geopandas(gpdpolygon)
    result = polygon.contains(point)
    assert gpdpolygon.contains(gpdpoint).values == result.values_host
    assert result.values_host[0] == expects
```

The third value in the `parametrize` statement is the value that is "expected", that is, the one that GeoPandas returns. The first thing to notice is that GeoPandas (GEOS) doesn't successfully identify these special cases both as being in the boundary, the first is a boundary case and the second is not.

I was able to identify EPSILON values to use for the `is_colinear` test and for the final intersection test (different epsilons are required because `is_colinear` needs to be squared to use as an inequality, whereas the intersection test is already an inequality so EPSILON can simply be added or subtracted.

After I identified working EPSILON values for these specific cases, I expanded them to higher precision examples:

```
        [Point([0.66, 0.006]), Polygon([[0, 0], [10, 1], [1, 1], [0, 0]]), False],
        [Point([0.666, 0.0006]), Polygon([[0, 0], [10, 1], [1, 1], [0, 0]]), False],
        [Point([3.33, 1.11]), Polygon([[6, 2], [3, 1], [3, 4], [6, 2]]), True],
        [Point([3.333, 1.111]), Polygon([[6, 2], [3, 1], [3, 4], [6, 2]]), True],
```

At this point GeoPandas's consistency breaks down. The first two work as expected. The third breaks my epsilon test, suggesting perhaps further EPSILON refinement. But the fourth example breaks GeoPandas's consistency: `[3.3, 1.1]` and `[3.33, 1.11` are "in" the polygon, but `[3.333, 1.111]` is not. It looks like GEOS uses a similar [RayCrossingCounter](https://github.com/libgeos/geos/blob/main/src/algorithm/RayCrossingCounter.cpp) with a complex object oriented design an no epsilons that I can see.

I'm inclined to drop epsilons and accept that boundary cases are fuzzy and prone to error. What do you think?

Reply:
@thomcom these are not corner cases. I'm saying that the approach is flawed. You can't compare directly to zero and expect consistent results. The examples I gave simply demonstrate that some decimal numbers are not representable in binary floating point. That was just an easy way to show the flaw in the `is_colinear` approach.

Reply:
Specifying Epsilon will probably just have different (fewer, probably) failure cases. The original algorithm also ignores epsilon as well for these kinds of cases and pip. I think I prefer leaving the epsilon out and just clarifying in the docs that floating point error may occur. I think it is implicit, just like with osGeo/GEOS having these kinds of cases in GeoPandas, just different ones than ours. I understand that strict equality is going to fail, doesn't everyone? :D @isVoid how do you feel about it after the discussion we had?

Reply:
> I'm inclined to drop epsilons and accept that boundary cases are fuzzy and prone to error. What do you think?

As long as this fits your need for DE-9IM, that's fine. I'm trying to help you avoid corner cases in the future where some geometry lies very close to a polygon and gets misclassified as `contains`/`intersects`.